File: src/Niche/Nim/chain_of_responsibility.nim
import std/strutils

type
  Handler = ref object
    name: string
    limit: int
    next: Handler

proc handle(handler: Handler, amount: int, visited: var seq[string]): string =
  visited.add(handler.name)
  if amount <= handler.limit:
    return "handled=" & handler.name & ";result=refund(" & $amount & ")"
  if handler.next.isNil:
    return "handled=none;result=rejected"
  handler.next.handle(amount, visited)

let escalation = Handler(name: "escalation", limit: high(int))
let billing = Handler(name: "billing", limit: 500, next: escalation)
let faq = Handler(name: "faq", limit: 50, next: billing)
var visited: seq[string] = @[]
let result = faq.handle(250, visited)
echo "visited=" & visited.join(">") & ";" & result

File: src/Web/Dart/chain_of_responsibility.dart
class Handler {
  Handler(this.name, this.limit, [this.next]);

  final String name;
  final int limit;
  final Handler? next;

  String handle(int amount, List<String> visited) {
    visited.add(name);
    if (amount <= limit) {
      return 'handled=$name;result=refund($amount)';
    }
    final successor = next;
    if (successor == null) {
      return 'handled=none;result=rejected';
    }
    return successor.handle(amount, visited);
  }
}

void main() {
  final escalation = Handler('escalation', 1 << 30);
  final billing = Handler('billing', 500, escalation);
  final faq = Handler('faq', 50, billing);
  final visited = <String>[];
  final result = faq.handle(250, visited);
  print('visited=${visited.join('>')};$result');
}
